lib/srv/reexec.go

```diff
@@ -1194,10 +1202,42 @@ func copyCommand(ctx *ServerContext, cmdmsg *ExecCommand) {
 	}
 }
 
-// CheckHomeDir checks if the user's home dir exists
+// HasAccessibleHomeDir checks if the given User has access to an existing home directory.
+func HasAccessibleHomeDir(localUser *user.User) error {
+	fi, err := os.Stat(localUser.HomeDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !fi.IsDir() {
+		return trace.NotFound("user home is not a directory")
+	}
+
+	stat, ok := fi.Sys().(*syscall.Stat_t)
+	if !ok {
+		return trace.AccessDenied("could not retrieve owner of home directory")
+	}
+
+	uid := strconv.Itoa(int(stat.Uid))
+	gid := strconv.Itoa(int(stat.Gid))
+
+	if uid == localUser.Uid && gid == localUser.Gid {
+		return nil
+	}
+
+	return trace.AccessDenied("user does not own configured home directory")
```

Comment:
This looks incorrect, here are some scenarios:
- if the home is owned by user `1000` and group `999`, but the user is `UID=1000`/`GID=1000` (`&&` vs `||`)
- user is a member of whatever group owning the home dir
- if home is `o+rwx`
- ACLs are used

Reply:
These are great points. We're probably better off not inspecting the the ownership information manually and instead just trying to stat the directory as the target user. This was mostly to avoid having to reexec if we didn't have to, but I think we _do_ have to :smile: I'll get a new commit up for this 👍